pkg/sql/logictest/testdata/logic_test/enums

```diff
@@ -418,7 +418,7 @@ SELECT ARRAY['hello']::_greeting, ARRAY['hello'::greeting]
 {hello} {hello}
 
 # Test that we can't mix enums in an array.
-query error pq: expected 'cockroach'::public.dbs to be of type greeting, found type dbs
+query error pq: expected 'cockroach'::dbs to be of type greeting, found type dbs
```

Comment:
This is a bit worrying to me -- if the prefix has gone away, that makes me worry that there is an `TypeReference` sitting around not resolved into a `types.T` anymore.

Reply:
I checked out the branch, and the test is fixed by this diff -- 
```
+++ b/pkg/sql/sem/tree/type_check.go
@@ -468,6 +468,7 @@ func (expr *CastExpr) TypeCheck(
        if err != nil {
                return nil, err
        }
+       expr.Type = exprType
        switch {
        case isConstant(expr.Expr):
                c := expr.Expr.(Constant)
```
It should have been done like this in the first place. We need to be setting `expr.Type` to the resolved types.T during type checking wherever expressions contain `TypeReferences`. This is so that they can be formatted correctly in different contexts -- the unresolved name doesn't give us any power. Your patch exposed the bug by returning early if the cast types were identical.